crates/ruff_linter/src/rules/pylint/rules/bad_string_format_type.rs

```diff
@@ -124,8 +124,8 @@ fn equivalent(format: &CFormatSpec, value: &Expr) -> bool {
         ResolvedPythonType::Atom(atom) => {
             // Special case where `%c` allows single character strings to be formatted
             if format.format_char == 'c' {
-                if let Expr::StringLiteral(string) = value {
-                    let mut chars = string.chars();
+                if let Expr::StringLiteral(ast::ExprStringLiteral { value, .. }) = value {
+                    let mut chars = value.chars();
```

Comment:
I suspect that you could implement `.chars()` without needing to allocate the string.

Reply:
Yes, we can.

Thinking about this with your previous comment, I think we should implement some other methods as well like `is_empty`, `len`, etc. which means that these APIs will guaranteed be free of allocation and, we would be providing `as_str` to get the concatenated source code to use methods not implemented on the abstract type.